linera-execution/solidity/linera.sol

```diff
@@ -18,16 +58,297 @@ library Linera {
+        uint32 multi_leader_rounds;
+        (new_pos, multi_leader_rounds) = bcs_deserialize_offset_uint32(new_pos, input);
+        bool open_multi_leader_rounds;
+        (new_pos, open_multi_leader_rounds) = bcs_deserialize_offset_bool(new_pos, input);
+        TimeoutConfig memory timeout_config;
+        (new_pos, timeout_config) = bcs_deserialize_offset_TimeoutConfig(new_pos, input);
+        return (
+            new_pos, ChainOwnership(super_owners, owners, multi_leader_rounds, open_multi_leader_rounds, timeout_config)
+        );
+    }
+
+    function bcs_deserialize_ChainOwnership(bytes memory input) public pure returns (ChainOwnership memory) {
+        uint256 new_pos;
+        ChainOwnership memory value;
+        (new_pos, value) = bcs_deserialize_offset_ChainOwnership(0, input);
+        require(new_pos == input.length, "incomplete deserialization");
+        return value;
+    }
+
+    function inner_chain_id(uint8 val) internal returns (bytes32) {
+        address precompile = address(0x0b);
+        bytes memory input1 = new bytes(2);
+        input1[0] = bytes1(uint8(0));
+        input1[1] = bytes1(val);
+        (bool success, bytes memory output1) = precompile.call(input1);
+        require(success);
+        return bcs_deserialize_bytes32(output1);
+    }
+
+    function chain_id() internal returns (bytes32) {
+        return inner_chain_id(uint8(0));
+    }
+
+    function application_creator_chain_id() internal returns (bytes32) {
+        return inner_chain_id(uint8(1));
+    }
+
+    function chain_ownership() internal returns (ChainOwnership memory) {
+        address precompile = address(0x0b);
+        bytes memory input1 = new bytes(2);
+        input1[0] = bytes1(uint8(0));
+        input1[1] = bytes1(uint8(2));
+        (bool success, bytes memory output1) = precompile.call(input1);
+        require(success);
+        return bcs_deserialize_ChainOwnership(output1);
+    }
+
+    function read_data_blob(bytes32 hash) internal returns (bytes memory) {
+        address precompile = address(0x0b);
+        bytes memory input1 = new bytes(2);
+        input1[0] = bytes1(uint8(0));
+        input1[1] = bytes1(uint8(3));
+        bytes memory input2 = abi.encodePacked(input1, hash);
+        (bool success, bytes memory output1) = precompile.call(input2);
+        require(success);
+        return output1;
+    }
+
+    function assert_data_blob_exists(bytes32 hash) internal {
+        address precompile = address(0x0b);
+        bytes memory input1 = new bytes(2);
+        input1[0] = bytes1(uint8(0));
+        input1[1] = bytes1(uint8(4));
+        bytes memory input2 = abi.encodePacked(input1, hash);
+        (bool success, bytes memory output1) = precompile.call(input2);
+        require(success);
+        assert(output1.length == 0);
+    }
+
+    struct OptionU32 {
+        bool has_value;
+        uint32 value;
```

Comment:
Is the code above generated or hand-written?

Reply:
Partially computer generated, partially hand written.

Reply:
Could we split them then? Manually-written from the generated ones and tag the generated file with an appropriate notice in the header about:
- how it was generated (command)
- when it was generated (timestamp)
- that it MUST NOT be modified manually (only via fixing the source generator)
?

Otherwise it's impossible to tell what should be generated and what has to be manually implemented.

For example, I don't know if I should review the functions that (de)serialize like bcs does b/c those _should_ be generated.

Reply:
Many changes were made to the generated code:
* I changed some types which were cumbersome and replaced them with "bytes32" or similar. And I want to change appropriately the `serde-reflection`.
* I also added checks for the nature of the object for boolean and also this will go into `serde-reflection`.
* I removed the functions that are unused.
* I reformatted the generated code with "forge fmt"
* Some types like `MessageIsBouncing` are a simpler version than the generated code.
* I avoided having `CryptoHash` and instead put directly "bytes32"
* Some name like "function chain_id" causes a namespace collision with other "chain_id" in the code that had to be renamed.

So, I am not sure how to document all that.

Reply:
I'd say that all of of the changes that you made to the generated code should be made to the generator itself. We should not modify the generated code. 

If you feel like there needs to be some "translation layer"/adapter between the generated code and this codebase then it should be added _in addition_ to the generated one AND in a separate file, with proper comments explaining why the generated output is not sufficient.

Reply:
Well, we need to have some contextual changes. How is the generated code supposed to know that we have a `function chain_id()`, and so that there is a namespace collision?

How is the `serde-generate` supposed to know which functions are used?

Otherwise, yes I agree that some of the changes will find its way into `serde-generate`. But that is not immediate.

Reply:
I didn't mean that the Linera API should end up in the generated code but all the mappings between Solidity<>Linera serialization could and should.

Reply:
In the end, the computer-generated types are not going to look great. Where is the harm in writing specific types?